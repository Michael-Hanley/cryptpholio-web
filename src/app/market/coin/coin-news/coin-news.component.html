<mat-list-item>
    <a href="{{url}}">
        <div class="list-item">
            <span><img class="article-image" src="{{imageUrl}}" /></span>
            <span class="article-title">{{title}}</span>
        </div>
    </a>
</mat-list-item>
<mat-divider></mat-divider>