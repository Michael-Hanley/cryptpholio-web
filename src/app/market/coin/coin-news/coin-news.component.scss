.article-image {
    margin: 5px 10px;
    max-width: 50px;
}
.article-title {
    margin-left: 10px;
}
