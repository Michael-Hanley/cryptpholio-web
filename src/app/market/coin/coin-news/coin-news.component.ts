import { Component, OnInit, Input } from '@angular/core';

@Component({
  selector: 'app-coin-news',
  templateUrl: './coin-news.component.html',
  styleUrls: ['./coin-news.component.scss']
})
export class CoinNewsComponent implements OnInit {
  @Input() title;
  @Input() imageUrl;
  @Input() url;
  constructor() { }

  ngOnInit() {
  }

}
