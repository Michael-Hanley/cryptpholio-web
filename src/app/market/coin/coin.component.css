.container {
    display: flex;
    /* flex-direction: column; */
    width: 100%;
    min-width: 350px;
    max-width: 1200px;
    overflow: auto;
    /* justify-content: center; */
}
.logo {
    /* display: inline;     */
    margin: 10px;
    /* float: left; */
}
.nameSymbol {
    margin-top: 15px;
}
.marketPrice {
    /* float: right; */
    margin-right: 10px;
    
}
.marketCap {
    margin-top: 15px;
    margin-right: 10px;    
    margin-left: auto;
    /* justify-content: flex-end; */
}
.icon {
    width: 12px;
    height: 12px;
}