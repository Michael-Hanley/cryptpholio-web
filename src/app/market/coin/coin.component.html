<div class="container" *ngIf="coin !== undefined">
    <div>
        <button mat-button (click)="backToMarker()">
            <mat-icon>arrow_back</mat-icon>Back to Market
        </button>
    </div>
    <br>
    <div class="coinInfo">
        <span class="logo"><img src="{{coin.icon}}" width="67" height="67"></span>
        <div>
            <div class="nameSymbol">
                <span>{{coin.name}}</span>
                <span>({{coin.symbol}})</span>
            </div>
            <div class="marketPrice">
                <div>{{coin.price | currency}}</div>
                <div>
                    <mat-icon class="icon" svgIcon="btc"></mat-icon>{{coin.priceBtc | number:'.2-9'}}
                </div>
            </div>
        </div>
        <div class='marketCap'>
            <div>Market Cap</div>
            <div>${{coin.marketCap | number: '1.0-0'}}</div>
            <div>
                <mat-icon class="icon" svgIcon="btc"></mat-icon>{{market_cap_btc  | number: '1.0-0'}}
            </div>
        </div>
    </div>
    <br>
    <div class='chart'>
        <app-main-chart 
            [name]='coin.name' 
            [symbol]='coin.symbol' 
            [supply]='coin.availableSupply'>
        </app-main-chart>
    </div>
    <ng-container *ngFor="let newsArticle of coinNews; let i=index">
        <app-coin-news *ngIf='maxNewsArticles(i)' 
            [title]='newsArticle.title' 
            [imageUrl]='newsArticle.imageurl'
            [url]='newsArticle.url'>
        </app-coin-news>
    </ng-container>
</div>
<div fxLayout fxLayoutAlign="center center">
    <mat-spinner class="spinner" *ngIf="coin === undefined">
    </mat-spinner>
</div>