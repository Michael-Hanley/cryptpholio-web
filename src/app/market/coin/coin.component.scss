.container {
    flex-direction: column;
    width: 100%;
    height: 100vh;
    justify-content: center;
}
.coinInfo {
  display: flex;
  flex-direction: row;
}
.logo {
    margin: 10px;
}
.nameSymbol {
    margin-top: 15px;
}
.marketPrice {
    margin-right: 10px;

}
.marketCap {
    margin-top: 15px;
    margin-right: 20px;
    margin-left: auto;
}
.icon {
    width: 12px;
    height: 12px;
}
