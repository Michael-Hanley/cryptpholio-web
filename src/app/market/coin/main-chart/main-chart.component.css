.spinner {
  margin-top: 125px;
  width: 350px;
  height: 350px;
}
.chartToggle {
  margin-left: 15px;
  margin-bottom: 1px;
}