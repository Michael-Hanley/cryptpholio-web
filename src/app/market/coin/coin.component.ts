import { Location } from '@angular/common';
import { Component, OnDestroy, OnInit } from '@angular/core';
import { MatIconRegistry } from '@angular/material';
import { DomSanitizer } from '@angular/platform-browser';
import { ActivatedRoute, Router } from '@angular/router';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { CoinService } from '../../services/coin.service';
import { NewsService } from '../../services/news.service';
import { RouteService } from '../../services/route.service';
import { Coin } from '../coin-market/coin-market.component';

type CoinDetailsReply = {
  coin: Coin
}

type SourceInfo = {
  name: String
  lang: String
  img: String
}

type CoinNews = {
  body: String
  categories: String
  downvotes: String
  guid: String
  id: String
  imageurl: String
  lang: String
  published_on: Number
  source: String
  source_info: SourceInfo
  tags: String
  title: String
  upvotes: String
  url: String
}

type CoinNewsReply = {
  Data: CoinNews[]
}

@Component({
  selector: 'app-coin',
  templateUrl: './coin.component.html',
  styleUrls: ['./coin.component.scss']
})
export class CoinComponent implements OnInit, OnDestroy {
  coin;
  current_btc_price;
  market_cap_btc;
  coinNews;
  imageUrl = 'https://www.cryptocompare.com';

  private ngUnsubscribe: Subject<any> = new Subject();

  constructor(private route: ActivatedRoute, iconRegistry: MatIconRegistry,
    public routeService: RouteService, sanitizer: DomSanitizer, public location: Location,
    private coinService: CoinService, private router: Router, private newsService: NewsService) {
    this.route.queryParams
      .pipe(takeUntil(this.ngUnsubscribe))
      .subscribe(params => {
        this.getCoinDetails(params.coinId);
      });
    iconRegistry.addSvgIcon(
      'btc',
      sanitizer.bypassSecurityTrustResourceUrl('assets/icons/btc.svg')
    );
  }

  getCoinDetails(coinId) {
    this.coinService.getCoinDetail(coinId)
      .pipe(takeUntil(this.ngUnsubscribe))
      .subscribe((coin: CoinDetailsReply) => {
        if (!coin) { return; }
        this.coin = coin.coin;
        this.current_btc_price = this.coin.price / this.coin.priceBtc;
        this.market_cap_btc = this.coin.marketCap / this.current_btc_price;
        this.getCoinNews(this.coin.symbol);
      });
  }

  getCoinNews(coin) {
    this.newsService.getCoinNews(coin)
      .pipe(takeUntil(this.ngUnsubscribe))
      .subscribe((res: CoinNewsReply) => {
        this.coinNews = res.Data;
      });
  }

  backToMarker() {
    this.router.navigate(['/market'], { queryParams: this.routeService.marketParams });
  }

  ngOnInit() { }
  ngOnDestroy(): any {
    this.ngUnsubscribe.next();
    this.ngUnsubscribe.complete();
  }
}
