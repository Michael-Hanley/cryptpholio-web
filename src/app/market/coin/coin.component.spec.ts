import { HttpClientTestingModule } from '@angular/common/http/testing';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { MatIconModule } from '@angular/material/icon';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { RouterTestingModule } from '@angular/router/testing';
import { MockComponent } from 'ng-mocks';
import { CoinNewsComponent } from './coin-news/coin-news.component';
import { CoinComponent } from './coin.component';
import { MainChartComponent } from './main-chart/main-chart.component';

let testProps = {
  coinNews: [{
    title: 'test-title',
    imageurl: 'test-image-url',
    url: 'test-url'
  }, {
    title: 'test-title-1',
    imageurl: 'test-image-url-1',
    url: 'test-url-1'
  }]
}

describe('CoinComponent', () => {
  let fixture: ComponentFixture<CoinComponent>;
  let component: CoinComponent;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [CoinComponent,
        MockComponent(CoinNewsComponent),
        MockComponent(MainChartComponent)],
      imports: [
        MatIconModule,
        MatProgressSpinnerModule,
        RouterTestingModule,
        HttpClientTestingModule
      ]
    })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CoinComponent);
    component = fixture.componentInstance;
    component.coinNews = testProps.coinNews;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
