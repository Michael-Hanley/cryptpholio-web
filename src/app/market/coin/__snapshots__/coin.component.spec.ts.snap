// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinComponent should match snapshot 1`] = `
<app-coin
  coinService={[Function CoinService]}
  location={[Function SpyLocation]}
  maxArticles={[Function Number]}
  newsService={[Function NewsService]}
  ngUnsubscribe={[Function Subject]}
  route={[Function ActivatedRoute]}
  routeService={[Function RouteService]}
  router={[Function Router]}
>
  <div
    fxlayout=""
    fxlayoutalign="center center"
  >
    
    <mat-spinner
      class="spinner mat-spinner mat-progress-spinner mat-primary mat-progress-spinner-indeterminate-animation"
      mode="indeterminate"
      role="progressbar"
      style="width: 100px; height: 100px;"
    >
      <svg
        focusable="false"
        preserveAspectRatio="xMidYMid meet"
        style="width: 100px; height: 100px;"
        viewBox="0 0 100 100"
      >
        
        <circle
          cx="50%"
          cy="50%"
          r="45"
          style="animation-name: mat-progress-spinner-stroke-rotate-100; stroke-dasharray: 282.7433388230814px; stroke-width: 10%;"
        />
        
      </svg>
    </mat-spinner>
  </div>
</app-coin>
`;
